kong/runloop/balancer.lua

```diff
@@ -473,8 +473,13 @@ do
   -- and upstream entity tables as values, or nil+error
   local function load_upstreams_dict_into_memory()
     local upstreams_dict = {}
-    for up in singletons.db.upstreams:each(1000) do
     -- build a dictionary, indexed by the upstream name
+    for up, err in singletons.db.upstreams:each(1000) do
+      if err then
+        log(CRIT, "could not obtain list of upstreams: ", err)
+        return upstreams_dict
```

Comment:
It may be more semantically correct to return `nil, err` here, since this is an mlcache callback. This way, we will avoid caching the result, and properly bubble up the error to `get_balancer()`, which will return `nil, err` as well, instead of `false`. Or was the return value of `get_balancer()` intentional? Currently, if other mlcache callback fail, they will make `get_balancer()` return `nil` and thus produce a 500 error.

Another approach would be to return `nil` here (no `err`), and thus respect a very short `neg_ttl` value (which is presently equal to `ttl` but we could override it for this call). Outside of the callback (maybe in `get_upstream_by_name()`), we can handle the `nil` return value appropriately.

Reply:
The idea that I get here is that we are attempting to fallback into a "no balancer" mode (as if no upstream was configured for this Service). Which is great, but since it is sub-optimal behaviour (for resiliency purposes), I feel like the second proposed approach is favourable.

The first approach only makes sense when we know there is an upstream, but we were just unable to fetch it (that said, we could be resilient and fallback to "no balancer" mode there as well, after all) :+1:

Reply:
You just read my mind: I was just going to ask whether it would be possible to not cache this to allow for retries or to set a ttl so that we could retry in case of an intermittent failure without hammering the database.

I pushed an updated version. What's a reasonable "very short `neg_ttl`"?

Reply:
> What's a reasonable "very short `neg_ttl`"?

Not a question that can be answered simply... tldr; 10s (the default you chose) seems reasonable. Otherwise, retry strategies play an important part in building resilient software (the C* driver has had constant and exponential retry strategies for a long time, but there are other Kong components, or first class citizens, that need it as well). Until then...